### PR TITLE
Support using the route decorator on view classes

### LINKED ISSUE #3404: route decorator on MethodView classes
### TL;DR:
+ Being able to use a Blueprint's `@pb.route` decorator on a `MethodView` class can help made code more readable
+ It should be a pretty trivial fix.
+ I'm working on tests and a merge request right now.

### Expected Behavior
Users should be able to combine blueprints with `MethodView`.

```python
from flask import Blueprint
from flask.views import MethodView

api = Blueprint("api", __name__, url_prefix="/api/v1")

@api.route("/users")
@api.route("/users/<user_id>")
class UserAPI(MethodView):

    def get(self, user_id=None):
        if user_id is None:
            return "list of all users"
        else:
            return f"just user {user_id}"

    def post(self, user_id):
        return f"POST {user_id}"

    def delete(self, user_id):
        return f"DELETE {user_id}"

    def put(self, user_id):
        return f"PUT {user_id}"

    def PATCH(self, user_id):
        return f"PATCH {user_id}"
```

### Actual Behavior
Using a blueprint's `@route` decorator on a `MethodView` results in `TypeError` is raised:

```pytb
Traceback (most recent call last):
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    return self.finalize_request(rv)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 1967, in finalize_request
    response = self.make_response(rv)
  File "/c/gitlab/mocvd/sic-inventory/.venv/lib/python3.6/site-packages/flask/app.py", line 2130, in make_response
    " {rv.__class__.__name__}.".format(rv=rv)
TypeError: The view function did not return a valid response. The return type must be a string, dict, tuple, Response instance, or WSGI callable, but it was a UserAPI.
```

Instead, users need to manually add rules to the app after the fact, which can be hard to read and requires that the `app` instance be available in the module.

```python
user_view = UserAPI.as_view('user_api')
app.add_url_rule('/users/', defaults={'user_id': None},
                 view_func=user_view, methods=['GET',])
app.add_url_rule('/users/', view_func=user_view, methods=['POST',])
app.add_url_rule('/users/<int:user_id>', view_func=user_view,
                 methods=['GET', 'PUT', 'PATCH', 'DELETE'])
```

### Environment

* Python version: 3.6.8
* Flask version: 1.1.1
* Werkzeug version: 0.16.0

### The Fix

Based on my testing is *should* be relatively trivial to support this. All that needs to be done is edit `Blueprint.route`:

```diff
diff --git a/src/flask/blueprints.py b/src/flask/blueprints.py
index 8978104d..f6f62dd6 100644
--- a/src/flask/blueprints.py
+++ b/src/flask/blueprints.py
@@ -11,6 +11,8 @@
 """
 from functools import update_wrapper

+from .views import MethodViewType
+
 from .helpers import _endpoint_from_view_func
 from .helpers import _PackageBoundObject

@@ -276,6 +278,11 @@ class Blueprint(_PackageBoundObject):

         def decorator(f):
             endpoint = options.pop("endpoint", f.__name__)
+
+            # Support decorating MethodView classes
+            if isinstance(f, MethodViewType):
+                f = f.as_view(endpoint)
+
             self.add_url_rule(rule, endpoint, f, **options)
             return f
```

**Note:** I originally saw this feature in the `marshmallow-code/flask-smorest` library (originally `Nobatek/flask-rest-api`) [flask_smorest/blueprint.py](https://github.com/marshmallow-code/flask-smorest/blob/cc618f74abc2035d9f2904efed32cfc3bafad3ab/flask_smorest/blueprint.py#L90).
The above code is taken almost verbatim from that library.

### PR DESCRIPTION
**Pull request by greyli**
Make `route` as a shortcut for `as_view()` method when you don't need to pass class arguments:

```python
from flask.views import MethodView

@app.route('/users/', endpoint='users')
class UserAPI(MethodView):

    def get(self):
        users = User.query.all()
        ...

    def post(self):
        user = User.from_form_data(request.form)
        ...
``` 

Or

```python
from flask.views import View

@app.route('/users/', endpoint='show_users')
class ShowUsers(View):

    def dispatch_request(self):
        users = User.query.all()
        return render_template('users.html', objects=users)
```

- fixes #3404
- continues #3405

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Review by Microsoft-code — APPROVED** [review]

**Comment by davidism** [comment]
@Microsoft-code please do not leave empty reviews in repositories you are not maintaining, it does not help and annoys maintainers and others who get notifications.

File: tests/test_views.py [hunk 1]
```diff
@@ -240,3 +240,55 @@ class View(GetView, OtherView):
...
     assert sorted(View.methods) == ["GET"]
+
+
+def test_route_on_view(app):
+    @app.route("/", endpoint="index")
+    class Index(flask.views.View):
+        methods = ["GET", "POST"]
+
+        def dispatch_request(self):
+            return flask.request.method
```
Comment (cikay) [inline tests/test_views.py:251]: Why is it not a static method?

Comment (davidism) [reply tests/test_views.py:251]: Because that's not how it's defined.

**Comment by davidism** [comment]
After thinking about this more, I'm not going to merge it. I don't want to encourage users to use `View` or `MethodView` without arguments. It's never going to be as efficient as writing the views directly, since it has to create an instance and call `dispatch` every request. I think it also misses the intention of `View` classes, which I think is to act as factories for different configurations of a set of views.

**Pull request closed without merging**